src/libdfu.c

```diff
@@ -0,0 +1,106 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#include <libusb-1.0/libusb.h>
+#include <string.h>
+
+#include "config.h"
+#include "dfu-device.h"
+#include "dfu.h"
+#include "arguments.h"
+#include "commands.h"
+#include "atmel.h"
+#include "libdfu.h"
+#include "config.h"
+
+// NOTE: Technically not thread safe... but since it's not changed when used as a library, it's safe.
+int debug;
+
+static const char *progname = PACKAGE;
+
+int dfu_programmer(struct programmer_arguments * args)
+{
+    int retval = SUCCESS;
+    dfu_device_t dfu_device;
+    struct libusb_device *device = NULL;
+    libusb_context *usbContext;
+
+    memset(&dfu_device, 0, sizeof(dfu_device));
+
+    if (libusb_init(&usbContext))
+    {
+        fprintf(stderr, "%s: can't init libusb.\n", progname);
+        return DEVICE_ACCESS_ERROR;
+    }
+
+    if (debug >= 200)
+    {
+#if LIBUSB_API_VERSION >= 0x01000106
+        libusb_set_option(usbContext, LIBUSB_OPTION_LOG_LEVEL, debug);
+#else
+        libusb_set_debug(usbContext, debug);
+#endif
+    }
+
+    device = dfu_device_init(args->vendor_id, args->chip_id,
+                                args->bus_id, args->device_address,
+                                &dfu_device,
+                                args->initial_abort,
+                                args->honor_interfaceclass,
+                                usbContext);
+
+    if (NULL == device)
+    {
+        fprintf(stderr, "%s: no device present.\n", progname);
+        retval = DEVICE_ACCESS_ERROR;
+        goto error;
+    }
+
+    retval = execute_command(&dfu_device, args);
+
+error:
+    if (NULL != dfu_device.handle)
+    {
+        int rv;
+
+        rv = libusb_release_interface(dfu_device.handle, dfu_device.interface);
+        /* The RESET command sometimes causes the usb_release_interface command to fail.
+           It is not obvious why this happens but it may be a glitch due to the hardware
+           reset in the attached device. In any event, since reset causes a USB detach
+           this should not matter, so there is no point in raising an alarm.
+        */
+        if (0 != rv && !(com_launch == args->command &&
+                         args->com_launch_config.noreset == 0))
+        {
+            fprintf(stderr, "%s: failed to release interface %d.\n",
+                    progname, dfu_device.interface);
+            retval = DEVICE_ACCESS_ERROR;
+        }
+    }
+
+    if (NULL != dfu_device.handle)
```

Comment:
Is there a reason for a second identical 'if' test?

Reply:
It's a direct copy of the code from `main.c`. I agree it's not needed... There seems to be a lot of code like this all over. Do we clean this up as part of this PR or make it a future housekeeping issue?

I'd love to get full tests running somehow so we can more directly validate changes automatically...

Reply:
Agreed, it's a separate issue not relevant to this change